spec/mongoid/contextual/mongo/persistence_context_spec.rb

```diff
@@ -0,0 +1,703 @@
+            end
+            expect(Band.count).to eq(0)
+          end
+
+          it "reads from the correct client" do
+            Band.with(client: :reports) do |klass|
+              expect(klass.count).to eq(1)
+              expect(klass.first).to eq(band)
+            end
+          end
+        end
+
+        context "when not using block form" do
+
+          let(:band) { Band.new.with(client: :reports) }
+
+          before do
+            band.save
+            expect(Band.count).to eq(0)
+          end
+
+          it "reads from the correct client" do
+            Band.with(client: :reports) do |klass|
+              expect(klass.count).to eq(1)
+              expect(klass.first).to eq(band)
+            end
+          end
+        end
+      end
+    end
+
+    context "when setting a client option" do
+
+      let(:client) { Band.collection.client }
+
+      before do
+        client.subscribe(Mongo::Monitoring::COMMAND, subscriber)
+      end
+
+      after do
+        client.unsubscribe(Mongo::Monitoring::COMMAND, subscriber)
+      end
+
+      let(:subscriber) do
+        EventSubscriber.new
+      end
+
+      context "when calling #with on a criteria" do
+
+        let(:criteria) do
+          Band.criteria
+        end
+
+        before do
+          5.times { Band.create! }
+        end
+
+        context "when using the block form" do
+
+          it "has the correct write preference" do
+            criteria.with(write: { w: 0 }) do |crit|
+              expect(crit.destroy_all).to eq(0)
+            end
+          end
+        end
+
+        context "when not using blocks" do
+
+          let!(:criteria) do
+            Band.criteria.with(write: { w: 0 })
+          end
+
+          it "has the correct write preference" do
+            expect(criteria.destroy_all).to eq(0)
+          end
+        end
+      end
+
+      context "when calling #with on the klass" do
+
+        before do
+          5.times { Band.create! }
+        end
+
+        context "when using the block form" do
+
+          it "has the correct write preference" do
+            Band.with(write: { w: 0 }) do |klass|
+              expect(klass.destroy_all).to eq(0)
+            end
+          end
+        end
+
+        context "when not using block form" do
+
+          let!(:criteria) { Band.with(write: { w: 0 }) }
+
+          it "has the correct write preference" do
+            expect(criteria.destroy_all).to eq(0)
+          end
+        end
+
+        context "when using an old criteria" do
+
+          let(:criteria) do
+            crit = Band.all
+            crit.view # load the view
+            crit.with(write: { w: 0 })
+          end
+
+          it "has the correct write preference" do
+            expect(criteria.destroy_all).to eq(0)
+          end
+        end
+      end
+
+      # TODO: cant figure out if this is relevant or not. havent been able to
+      # test effectively
+      xcontext "when calling #with on a document" do
```

Comment:
Any ideas on how to test this? this is for testing client options (like read preference or write concern) when working with a single document....

Reply:
How about `document.with(read: { mode: :primary }).reload`, ensure reload does a query to primary?